R/tax_unique.R

```diff
@@ -0,0 +1,279 @@
+#' Filter occurrences to unique taxa
+#'
+#' A function to filter a list of taxonomic occurrences to unique taxa of a
+#' predefined resolution. In many cases palaeobiologists achieve this by
+#' retaining only unique occurrences identified to the given taxonomic
+#' resolution, however here we also retain occurrences identified to a coarser
+#' resolution which are not already represented within the dataset.
+#' If we take the following set of occurrences:
+#' Albertosaurus sarcophagus
+#' Ankylosaurus sp.
+#' Aves indet.
+#' Ceratopsidae indet.
+#' Hadrosauridae indet.
+#' Ornithominus sp.
+#' Tyrannosaurus rex
+#' then a filter for species-level identifications will reduce the species
+#' diversity to two. However, none of these clades are nested within another, so
+#' each of the indeterminately identified occurrences represents at least one
+#' species not already represented in the dataset. This function is designed to
+#' deal with such taxonomic data, and would retain all seven 'species' in this
+#' dataset.
+#' Data can be supplied in one of two forms, either a dataframe containing
+#' standard occurrence data from the Paleobiology Database, or a series of
+#' vectors which each describe a different taxonomic level for a set of
+#' occurrences (the order of names in the vectors must correspond). At a
+#' minimum, species, genus, and family vectors must be supplied. Occurrence
+#' data can be filtered to retain either unique species or unique genera, and
+#' can be defined as unique either across the whole dataset or within set
+#' categories (e.g. temporal intervals, spatial bins). Missing data should be
+#' indicated with NAs.
+#'
+#' @param paleobioDB \code{dataframe}. A dataframe of taxonomic occurrences
+#' downloaded directly from the Paleobiology Database. The dataframe must
+#' include the following columns: class, order, family, genus, accepted_name.
+#' @param species \code{character}. A vector of species names.
+#' @param genus \code{character}. A vector of genus names.
+#' @param family \code{character}. A vector of family names.
+#' @param order \code{character}. A vector of order names.
+#' @param class \code{character}. A vector of class names.
+#' @param resolution \code{character}. The taxonomic resolution at which to
+#' identify unique occurrences, either species (the default) or genera.
+#' @param by \code{numeric, character}. A category within which to determine
+#' unique taxa. When using \code{paleobioDB}, this should specify the
+#' column within the dataframe to use as the category. When using individual
+#' taxonomic vectors, this should be an additional vector of numbers or
+#' character-based labels designating which category each occurrence belongs to.
+#'
+#' @return A \code{dataframe} of unique taxa, with row numbers corresponding to
```

Comment:
I guess all input occurrences would be unique anyway? At least with PBDB data. I haven't come across an instance of the same occurrence repeated in a dataset but with different taxonomic names. I could see it happening... but this would be quite difficult to identify? Maybe (likely) I am misunderstanding your point Will...

Reply:
I just thought users might want the rest of the occurrence information returned, but I guess the point here is to get unique taxa not unique occurrences, so you really only want the taxonomic information returned.

Reply:
No no, I agree. I think the entire input occdf should be returned with an additional column providing respective unique names... I just wasn't sure what you meant by a 'reduced' occdf.

Reply:
If there are unique names, then there must be non-unique names, right? So can we filter those rows out of the occdf?

Reply:
I would opt for keeping all original data, and just supplying an additional column with the unique names... this would be more useful for working with diversity analysis in my opinion, but maybe you have other analysis in mind? I'm imagining a situation in which you have duplicated names but there is additional information in the occdf (location, age, etc). They are unique occurrences, but maybe not unique taxa. Not sure this is info we want to discard?

Reply:
Sure, that's fine with me!

Reply:
I don't really understand this. The current output filters the input dataset to one row per unique 'species' (admittedly this needs to be better explained in the documentation), so the number of rows can be counted to provide the diversity estimate. The current output provides the full taxonomic breakdown, which would be quite complicated to collapse into a single variable. The other problem with attaching this information to the input dataframe is that the way the code currently works is that the individual occurrences retained are arbitrary: if two occurrences are present in the dataframe for "Diplodocidae indet.", then the first (highest) will be retained and the others will be filtered out. Would these repetitions require NAs, or just the same name repeated? This could become confusing - I would advocate for diversity being calculated using just the unique names, whereas this approach would also provide frequency data. Doing all of this is probably possible but is quite a long way from what I envisioned / have already implemented, so I want to make sure I understand what we want before trying to do this.

Reply:
I suppose if someone really cares about a particular column of data, they can just use that as the grouping column, so the new format is fine with me (although see my comment above about adding a column for the "unique" names).